discovery: replies should only contain a subset of the requested

lazan opened the pull request:
Before: when we fan out the replies for accumulated discovery replies from other domains we do not filter base on the requested services and queues. Hence, the replies to caller could contain services and queues that was not asked for. Which lead to domains knows about services/queues that there hasn't been a need for.

Now: The accumulated reply content is intersected with what the actual caller has requested -> a subset of the requested is replied to caller.

lazan committed:
discovery: replies should only contain a subset of the requested

Before: when we fan out the replies for accumulated discovery replies
from other domains we do not filter base on the requested services and
queues. Hence, the replies to caller could contain services and queues
that was not asked for. Which lead to domains knows about
services/queues that there hasn't been a need for.

Now: The accumulated reply content is intersected with what the actual
caller has requested -> a subset of the requested is replied to caller.

ahhud commented on middleware/domain/include/domain/discovery/state.h:
```diff
@@ -356,6 +356,68 @@ namespace casual
+                  //! adds all cached resources to services and queues
+                  request_content complement( request_content&& content);
+
+                  //! @returns the "cached" requested resources associated withe the correlation intersected with the 
```
spelling `withe`

ahhud commented on middleware/domain/source/discovery/handle.cpp:
```diff
@@ -226,25 +218,24 @@ namespace casual
                         for( auto& destination : discoveries)
                         {
+                           // filter based on the requested resources.
+                           reply.content = state.in_flight_cache.filter_reply( destination.correlation, content);
```
Nice!

ahhud commented on middleware/domain/source/discovery/handle.cpp:
```diff
@@ -497,6 +485,8 @@ namespace casual
                      return;
                   }
 
+                  state.in_flight_cache.add( message);
```
In case of a local lookup path in below statements, I think there is a obsolete cache registration never removed.

lazan replied:
Yeah, I think you're right.

ahhud commented on middleware/domain/source/discovery/state.cpp:
```diff
@@ -228,19 +228,115 @@ namespace casual
+                     for( auto& pair : resources)
+                     {
+                        // clean upp total "cached" services
+                        if( auto found = algorithm::find( m_resource_count, pair.second); --found->second == 0)
```
Personally I think this type of compound statement is somewhat hard to read and grasp. But that is just because I am not used to it. :-)

lazan replied:
You mean with _if init statement_ in general?

lazan replied:
In some cases you can narrow the scope of a variable to the if scope with _init-statement_, which often is worth it. In this particular case it could be done with two nested if statements. Which is ok if it's one-liners, I think.

ahhud replied:
If there is a very long init statement the actual condition appear at the very far right. 
Perhaps a different format is clearer. I.e 
if( init...;
     condition)
   statement;

But it is only a matter of taste.

ahhud reviewed (changes requested):
Its really only the possible obsolete cache registration which needs addressing. Rest is more subjective. :-)

lazan committed:
discovery: "local" known services/queues are not added to discovery reply

Before: Scenrio
* inbound discovery request for x, y
* x is "locally" known
* y is not known, an external-forward-request is sent.
* When the reply of the external-request is received the "locally" known
  is "lost", and not added to the reply

Now: We cache the "locally" known and union it whith the external reply
and reply this to the caller.
